### PR TITLE
Remove ISchema in GroupTransform

### PR DESCRIPTION
**Pull request by wschin**
Another ISchema implementation is going to leave. This PR is part of #1501.

### CODE REVIEW

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 1]
```diff
@@ -180,35 +180,42 @@ public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Ra
         }
 
         /// <summary>
+        /// The output schema collects columns used to group columns (aka group-key columns) columns being grouped from the source scheam.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >columns (aka group-key columns) columns [](start = 61, length = 39)

columns columns columns!
I think you missed columns between columns and columns.
Seriously, i think last columns is not necessary.

Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >scheam [](start = 131, length = 6)

tyop

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 2]
```diff
@@ -180,35 +180,42 @@ public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Ra
...
+        /// In an output schema, group-key columns are followed by grouped columns. For example, if column "Age" is used to group "UserId" column,
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >grouped [](start = 67, length = 7)

I would call them aggregated columns.

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 1 above)
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >The output schema collects [](start = 12, length = 26)

Output schema doesn't do anything right, it just contains information. To claim it collects something is wrong. You can state what it's created by some actions, but I don't think it's right to say it's doing actions by itself.

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 3]
```diff
@@ -180,35 +180,42 @@ public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Ra
...
         /// For keep columns, the type is Vector of original type and variable length.
         /// The only metadata preserved is the KeyNames and IsNormalized.
         ///
         /// All other columns are dropped.
         /// </summary>
-        private sealed class GroupSchema : ISchema
+        private sealed class GroupBinding
         {
             private static readonly string[] _preservedMetadata =
                 new[] { MetadataUtils.Kinds.IsNormalized, MetadataUtils.Kinds.KeyValues };
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: either use it in BuildOutputSchema or delete it.

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 2 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Sure.

---
In reply to: [245808905](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245808905) [](ancestors = 245808905)

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Ok.

---
In reply to: [245809414](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245809414) [](ancestors = 245809414)

Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: I will rewrite the whole summary.

---
In reply to: [245808197](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245808197) [](ancestors = 245808197)

Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Nice catch.

---
In reply to: [245808223](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245808223) [](ancestors = 245808223)

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Deleted.

---
In reply to: [245810211](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245810211) [](ancestors = 245810211)

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 4]
```diff
@@ -180,35 +180,36 @@ public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Ra
...
             private readonly string[] _keepColumns;
 
-            public readonly int[] GroupIds;
-            public readonly int[] KeepIds;
-
-            private readonly int _groupCount;
-            private readonly ColumnType[] _columnTypes;
-
-            private readonly Dictionary<string, int> _columnNameMap;
+            // GroupIds[i] is the i-th group(-key) column's column index in the source schema.
```
Comment (singlis) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >GroupIds [](start = 15, length = 8)

GroupIds looks to be renamed to GroupColumnIndexes - should this comment be updated?

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 5]
```diff
@@ -180,35 +180,36 @@ public override RowCursor[] GetRowCursorSet(Func<int, bool> predicate, int n, Ra
...
+            public readonly int[] GroupColumnIndexes;
+            // GroupIds[i] is the i-th aggregated column's column index in the source schema.
```
Comment (singlis) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >GroupIds [](start = 15, length = 8)

KeepColumnIndexes?

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 6]
```diff
@@ -244,52 +242,62 @@ public GroupSchema(Schema inputSchema, IHostEnvironment env, ModelLoadContext ct
...
-                    var primitiveType = srcType as PrimitiveType;
-                    Contracts.Assert(primitiveType != null);
-                    types[i] = new VectorType(primitiveType, size: 0);
+                    // Prepare column's metadata.
+                    var metadataBuilder = new MetadataBuilder();
+                    metadataBuilder.Add(sourceSchema[groupValueColumnName].Metadata,
+                        s => s == MetadataUtils.Kinds.IsNormalized || s == MetadataUtils.Kinds.KeyValues);
+
+                    // Prepare column's type.
+                    var aggragatedValueType = sourceSchema[groupValueColumnName].Type as PrimitiveType;
```
Comment (singlis) [inline src/Microsoft.ML.Transforms/GroupTransform.cs]: >aggragatedValueType [](start = 24, length = 19)

spelling - should be aggregatedValueType

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 5 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Fixed. Thanks.

---
In reply to: [245831904](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245831904) [](ancestors = 245831904)

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 4 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Right. Thank you.

---
In reply to: [245831781](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245831781) [](ancestors = 245831781)

File: src/Microsoft.ML.Transforms/GroupTransform.cs (hunk 6 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Nice catch!

---
In reply to: [245832427](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245832427) [](ancestors = 245832427)

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 7]
```diff
@@ -319,103 +327,47 @@ public void Save(ModelSaveContext ctx)
...
                 for (int i = 0; i < names.Length; i++)
                 {
-                    int col;
-                    if (!schema.TryGetColumnIndex(names[i], out col))
+                    // Find column called names[i] from input schema.
+                    var retrievedColumn = schema.GetColumnOrNull(names[i]);
+
+                    // Throw if no such a column in schema.
+                    if (!retrievedColumn.HasValue)
                         throw except(string.Format("Could not find column '{0}'", names[i]));
```
[the remark this answers is not in the document]
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: It switches to 
```
var errorMessage = string.Format("Could not find column '{0}'", names[i]);
_ectx.Check(retrievedColumn.HasValue, errorMessage);
```

---
In reply to: [245833518](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245833518) [](ancestors = 245833518)

[the remark this answers is not in the document]
Comment (singlis) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: what about string.Format($"Could not find column '{names[i]}'");?

---
In reply to: [245836894](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245836894) [](ancestors = 245836894,245833518)

[the remark this answers is not in the document]
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: I hope you mean just `$"Could not find column '{names[i]}'"` `string.Format` isn't necessary if you use `$"something"`

---
In reply to: [245845003](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245845003) [](ancestors = 245845003,245836894,245833518)

[the remark this answers is not in the document]
Comment (singlis) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: yes I do

---
In reply to: [245845630](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245845630) [](ancestors = 245845630,245845003,245836894,245833518)

File: src/Microsoft.ML.Transforms/GroupTransform.cs [hunk 8]
```diff
@@ -242,54 +247,64 @@ public GroupSchema(Schema inputSchema, IHostEnvironment env, ModelLoadContext ct
...
-                    Contracts.Assert(primitiveType != null);
-                    types[i] = new VectorType(primitiveType, size: 0);
+                    // Prepare column's metadata.
+                    var metadataBuilder = new MetadataBuilder();
+                    metadataBuilder.Add(sourceSchema[groupValueColumnName].Metadata,
+                        s => s == MetadataUtils.Kinds.IsNormalized || s == MetadataUtils.Kinds.KeyValues);
+
+                    // Prepare column's type.
+                    var aggregatedValueType = sourceSchema[groupValueColumnName].Type as PrimitiveType;
+                    if (aggregatedValueType == null)
```
[the remark this answers is not in the document]
Comment (wschin) [reply src/Microsoft.ML.Transforms/GroupTransform.cs]: Totally agree with you.

---
In reply to: [245845559](https://github.com/dotnet/machinelearning/pull/2060#discussion_r245845559) [](ancestors = 245845559)

**Pull request merged**